File: src/runtime/typescript/src/mesh-job-submitter.ts
/**
 * Consumer-side `MeshJob` submitter (Phase 1 — MeshJob substrate).
 *
 * Mirrors Python's `_mcp_mesh.engine.mesh_job_submitter.MeshJobSubmitter`.
 *
 * When a consumer tool declares a parameter typed as `MeshJob` (i.e.
 * `meshJobParamIndex` is set on the resolved tool definition), the
 * runtime injects an instance of this class into that slot. The user
 * calls `await submitter.submit({ payload: ..., maxDuration: ... })` to
 * enqueue work on the remote producer and gets back a `JobProxy`
 * exposing `wait`/`status`/`cancel`.
 *
 * Per the design's "decouple resolver from capability metadata"
 * decision: this class does NOT verify the target capability is
 * registered with `task=true`. The registry rejects `submit_job`
 * against a non-task capability with a clear error which surfaces
 * verbatim to the caller.
 *
 * Resilience: `submit` retries on transient registry errors (network
 * drops / 5xx / 503) up to 3 attempts with 200ms then 1s backoff
 * between them. 4xx / NotFound / Conflict / serialization errors
 * propagate immediately — those won't self-heal.
 */
import { submitJob, type JobProxy } from "@mcpmesh/core";

const _SUBMIT_MAX_ATTEMPTS = 3;
// Backoff applied BETWEEN attempts. With 3 attempts there are 2 gaps,
// so this array has exactly 2 entries — a third entry would never be
// reached (the previous `[200, 1000, 5000]` had the 5s tier silently
// dead). Tests assert 3 total attempts; mirror that here precisely.
const _SUBMIT_BACKOFF_MS = [200, 1000];

/**
 * Heuristic: does this error string represent a transient registry
 * failure that's worth retrying? Errs on the side of NOT retrying when
 * the signal is ambiguous — submit failures are user-facing and a fast
 * clear failure beats a slow flaky one.
 *
 * The Rust napi binding propagates `JobError` as `Error(message)`. The
 * Rust side emits messages like:
 *   "backend error: network error: <reqwest::Error>"
 *   "backend error: backend unavailable: <body>"
 *   "backend error: server error (HTTP 502): <body>"
 *   "backend error: backend error: HTTP 400: <body>"   (4xx, NOT transient)
 *   "backend error: job not found: <id>"               (404, NOT transient)
 *   "backend error: conflict: <reason>"                (409, NOT transient)
 */
function isTransientSubmitError(err: unknown): boolean {
  const msg = String((err as Error)?.message ?? err).toLowerCase();
  if (msg.includes("network error")) return true;
  if (msg.includes("backend unavailable")) return true;
  if (msg.includes("server error (http 5")) return true;
  return false;
}

/**
 * Per-submission options for `MeshJobSubmitter.submit(...)`. Matches
 * the napi `SubmitJobArgs` shape (minus `registryUrl`/`capability`/
 * `submittedBy` which the submitter binds at construction).
 */
export interface SubmitOptions {
  /** Per-attempt soft timeout (seconds). `undefined` ≡ registry default. */
  maxDuration?: number;
  /**
   * Maximum retries on failure (registry default is 1 — one attempt,
   * no retry). `undefined` ≡ registry default.
   */
  maxRetries?: number;
  /**
   * Absolute wall-clock deadline across all retries. Accepts a `Date`
   * or a unix-epoch number (seconds). `undefined` ≡ unlimited per the
   * "Resolved Decisions" section of MESHJOB_DESIGN.org.
   */
  totalDeadline?: Date | number;
}

/**
 * Per-binding submit handle for a `MeshJob`-typed dependency.
 *
 * Constructed by the dependency injector at agent startup, one per
 * (consumer-tool, dependency-capability) pair.
 */
export class MeshJobSubmitter {
  /** Remote capability this submitter targets. */
  readonly capability: string;
  /** Identifier written to the registry's `submitted_by` column. */
  readonly submittedBy: string;
  /** Base URL of the mesh registry. */
  readonly registryUrl: string;

  constructor(
    capability: string,
    submittedBy: string,
    registryUrl: string,
  ) {
    this.capability = capability;
    this.submittedBy = submittedBy;
    this.registryUrl = registryUrl;
  }

  /**
   * Submit a new job on the bound capability and return a `JobProxy`.
   *
   * The first positional argument carries the user-supplied payload
   * fields the producer will receive as its tool args. Optional
   * second-arg knobs map to registry columns (`max_duration`,
   * `max_retries`, `total_deadline`).
   *
   * @example
   * ```ts
   * const proxy = await jobSubmitter.submit(
   *   { user_id: "demo", sections: ["intro", "summary"] },
   *   { maxDuration: 60 },
   * );
   * const result = await proxy.wait(60);
   * ```
   *
   * Error semantics:
   *   - Transient registry errors (network/5xx/503) → up to 3 attempts
   *     with 200ms then 1s backoff between them before propagating.
   *   - 4xx / NotFound / Conflict / serialization → fail-fast.
   */
  async submit(
    payload: Record<string, unknown>,
    options: SubmitOptions = {},
  ): Promise<JobProxy> {
    let totalDeadlineEpoch: number | undefined;
    if (options.totalDeadline !== undefined) {
      if (options.totalDeadline instanceof Date) {
        totalDeadlineEpoch = Math.floor(options.totalDeadline.getTime() / 1000);
      } else if (typeof options.totalDeadline === "number") {
        // If looks like ms, downgrade to seconds. Anything > 1e12 is ms.
        totalDeadlineEpoch =
          options.totalDeadline > 1e12
            ? Math.floor(options.totalDeadline / 1000)
            : Math.floor(options.totalDeadline);
      } else {
        throw new TypeError(
          `MeshJobSubmitter.submit: totalDeadline must be a Date or a number ` +
            `(unix epoch seconds); got ${typeof options.totalDeadline}`,
        );
      }
    }

    let lastErr: unknown = null;
    for (let attempt = 1; attempt <= _SUBMIT_MAX_ATTEMPTS; attempt++) {
      try {
        const proxy = await submitJob({
          registryUrl: this.registryUrl,
          capability: this.capability,
          payload,
          submittedBy: this.submittedBy,
          // Let the registry assign owner via pull-claim.
          ownerInstanceId: undefined,
          maxDuration: options.maxDuration,
          maxRetries: options.maxRetries,
          totalDeadline: totalDeadlineEpoch,
        });
        return proxy;
      } catch (err) {
        lastErr = err;
        if (!isTransientSubmitError(err) || attempt >= _SUBMIT_MAX_ATTEMPTS) {
          throw err;
        }
        const backoffMs =
          _SUBMIT_BACKOFF_MS[Math.min(attempt - 1, _SUBMIT_BACKOFF_MS.length - 1)];
        await new Promise((resolve) => setTimeout(resolve, backoffMs));
      }
    }
    // Defensive — unreachable: we either return or throw above.
    throw lastErr ?? new Error("MeshJobSubmitter.submit: unreachable");
  }
}

File: src/runtime/python/_mcp_mesh/engine/mesh_job_submitter.py
"""Consumer-side ``MeshJob`` submitter (Phase 1 — MeshJob substrate).

When a consumer tool declares a parameter annotated ``MeshJob`` whose name
matches a declared dependency capability, the runtime injects an instance
of :class:`MeshJobSubmitter` into that slot. The user calls
``await submitter.submit(...)`` to enqueue work on the remote producer
and gets back a :class:`mcp_mesh_core.JobProxy` to await/poll/cancel.

This is a thin Python wrapper around :func:`mcp_mesh_core.submit_job` —
it carries the per-binding context (capability + submitted_by +
registry_url) so the user only supplies the call payload + per-submission
knobs (``max_duration``, ``max_retries``, ``total_deadline``).

Per the design's "decouple resolver from capability metadata" decision:
this class does NOT verify that the target capability is registered with
``task=True`` — the registry rejects ``submit_job`` against a
non-task capability with a clear error, and that error surfaces to the
caller verbatim. Doing the check here would require pulling capability
metadata into the DDDI resolver (which currently knows only types).

Resilience: ``submit`` retries on transient registry errors (network
drops, 5xx, 503) up to a small bounded number of attempts. 4xx /
serialisation / NotFound errors propagate immediately — those won't
self-heal so retrying just delays the user-facing failure.
"""

from __future__ import annotations

import asyncio
import logging
from datetime import datetime
from typing import Any, Optional

logger = logging.getLogger(__name__)


# ---------------------------------------------------------------------------
# Transient-error retry helper
# ---------------------------------------------------------------------------
#
# Rust exposes JobError to Python only as RuntimeError(message), so we
# fall back to substring checks against the error string. The Rust side
# emits messages like:
#   "backend error: network error: <reqwest::Error>"
#   "backend error: backend unavailable: <body>"
#   "backend error: server error (HTTP 502): <body>"
#   "backend error: backend error: HTTP 400: <body>"   (4xx, NOT transient)
#   "backend error: job not found: <id>"               (404, NOT transient)
#   "backend error: conflict: <reason>"                (409, NOT transient)
# We classify by these stable substrings.

_SUBMIT_MAX_ATTEMPTS = 3
# Backoff per attempt (seconds): 200ms, 1s, 5s. Last slot unused (we
# only sleep BETWEEN attempts so N-1 entries are consulted).
_SUBMIT_BACKOFF_SECS = (0.2, 1.0, 5.0)


def _is_transient_submit_error(exc: BaseException) -> bool:
    """Heuristic: does this RuntimeError represent a transient registry
    failure that's worth retrying? Errs on the side of NOT retrying when
    the signal is ambiguous — submit failures are user-facing and a
    fast clear failure beats a slow flaky one."""
    msg = str(exc).lower()
    # Definite transients (Rust BackendError::is_transient == true).
    if "network error" in msg:
        return True
    if "backend unavailable" in msg:
        return True
    if "server error (http 5" in msg:
        return True
    # Defensively don't retry on 4xx / NotFound / Conflict / serialisation.
    return False


class MeshJobSubmitter:
    """Per-binding submit handle for a ``MeshJob``-typed dependency.

    Constructed by the dependency injector at call time, one per
    (consumer-function, dependency-capability) pair.

    Attributes:
        capability: The remote capability this submitter targets (matches
            the dependency declaration on the consumer's ``@mesh.tool``).
        submitted_by: Identifier written to the registry's
            ``submitted_by`` column. Defaults to the running agent's
            instance id, so the registry can correlate this submission
            with the calling agent.
        registry_url: Base URL of the mesh registry. Reused for every
            submission via this submitter.
    """

    def __init__(
        self,
        capability: str,
        submitted_by: str,
        registry_url: str,
    ) -> None:
        self.capability = capability
        self.submitted_by = submitted_by
        self.registry_url = registry_url

    def __repr__(self) -> str:  # pragma: no cover - diagnostic
        return (
            f"MeshJobSubmitter(capability={self.capability!r}, "
            f"submitted_by={self.submitted_by!r})"
        )

    async def submit(
        self,
        *,
        max_duration: Optional[int] = None,
        max_retries: Optional[int] = None,
        total_deadline: Optional[datetime] = None,
        **payload: Any,
    ) -> Any:
        """Submit a new job on the bound capability and return a
        :class:`mcp_mesh_core.JobProxy`.

        Args:
            max_duration: Per-attempt soft timeout (seconds). ``None`` ≡
                registry default (300s, see design doc).
            max_retries: Maximum retries on failure. ``None`` ≡ registry
                default (1 attempt — no retry).
            total_deadline: Absolute wall-clock deadline across all
                retries. ``None`` ≡ unlimited (per "Resolved Decisions"
                in design doc). When supplied, converted to a Unix-epoch
                integer for the registry.
            **payload: User-supplied call arguments — passed verbatim
                as ``submitted_payload`` on the new job row. Must be
                JSON-serialisable (MCP tool results / args are JSON by
                spec).

        Returns:
            :class:`mcp_mesh_core.JobProxy` bound to the new job's id.
            Use ``await proxy.wait(timeout_secs=...)`` to await
            terminal status, or ``await proxy.status()`` for a single
            registry read.

        Raises:
            RuntimeError: If the registry rejects the submission (e.g.
                target capability isn't registered with ``task=True``,
                or the registry is unreachable). The underlying error
                message is propagated for diagnosability.
        """
        # Lazy import: keeps the SDK importable in test environments
        # where the native mcp_mesh_core extension isn't available.
        try:
            from mcp_mesh_core import submit_job
        except Exception as e:
            raise RuntimeError(
                f"MeshJobSubmitter.submit: mcp_mesh_core.submit_job "
                f"unavailable ({e}); cannot submit job to capability "
                f"{self.capability!r}"
            ) from e

        # Normalise total_deadline → Unix epoch int (the schema column is
        # INTEGER per MESHJOB_DESIGN.org "Schema").
        total_deadline_epoch: Optional[int] = None
        if total_deadline is not None:
            try:
                total_deadline_epoch = int(total_deadline.timestamp())
            except Exception as e:
                raise ValueError(
                    f"MeshJobSubmitter.submit: total_deadline must be a "
                    f"datetime (got {type(total_deadline).__name__}): {e}"
                ) from e

        logger.debug(
            "MeshJobSubmitter.submit: capability=%s submitted_by=%s "
            "max_duration=%s max_retries=%s payload_keys=%s",
            self.capability,
            self.submitted_by,
            max_duration,
            max_retries,
            list(payload.keys()),
        )

        # Bounded retry on transient registry errors (network blips, 5xx,
        # 503). 4xx / NotFound / Conflict propagate immediately — they
        # won't self-heal in 5 seconds. Submit failures are user-facing
        # so we keep the window small (3 attempts, max ~6.2s total).
        last_exc: Optional[BaseException] = None
        for attempt in range(1, _SUBMIT_MAX_ATTEMPTS + 1):
            try:
                proxy = await submit_job(
                    self.registry_url,
                    self.capability,
                    payload,
                    self.submitted_by,
                    None,  # owner_instance_id — let the registry pull-claim assign it
                    max_duration,
                    max_retries,
                    total_deadline_epoch,
                )
                if attempt > 1:
                    logger.info(
                        "MeshJobSubmitter.submit recovered on attempt %d "
                        "(capability=%s)",
                        attempt,
                        self.capability,
                    )
                return proxy
            except RuntimeError as e:
                last_exc = e
                if not _is_transient_submit_error(e):
                    # 4xx / NotFound / Conflict / serialisation — fail fast.
                    raise
                if attempt >= _SUBMIT_MAX_ATTEMPTS:
                    logger.warning(
                        "MeshJobSubmitter.submit giving up after %d attempts "
                        "(capability=%s, last_err=%s)",
                        attempt,
                        self.capability,
                        e,
                    )
                    raise
                # Backoff index uses attempt-1 (so attempt 1 sleeps 0.2s).
                backoff = _SUBMIT_BACKOFF_SECS[
                    min(attempt - 1, len(_SUBMIT_BACKOFF_SECS) - 1)
                ]
                logger.info(
                    "MeshJobSubmitter.submit transient failure on attempt %d "
                    "(capability=%s, err=%s); retrying in %.2fs",
                    attempt,
                    self.capability,
                    e,
                    backoff,
                )
                await asyncio.sleep(backoff)
        # Defensive — unreachable: we either return a proxy or raise above.
        assert last_exc is not None
        raise last_exc
